types/tipset.go

```diff
@@ -2,155 +2,148 @@ package types
+			cmp = bytes.Compare(sorted[i].Cid().Bytes(), sorted[j].Cid().Bytes())
+		}
+		return cmp < 0
+	})
 
-	id := b.Cid()
-	ts[id] = b
-	return nil
+	return TipSet{sorted}, nil
 }
 
-// Clone returns a shallow copy of the TipSet.
-func (ts TipSet) Clone() TipSet {
-	r := TipSet{}
-	for k, v := range ts {
-		r[k] = v
-	}
+// Defined checks whether the tipset is defined.
+// Invoking any other methods on an undefined tipset will result in undefined behaviour (c.f. cid.Undef)
+func (ts TipSet) Defined() bool {
+	return len(ts.blocks) > 0
+}
 
-	return r
+// Len returns the number of blocks in the tipset.
+func (ts TipSet) Len() int {
+	return len(ts.blocks)
 }
 
-// String returns a formatted string of the TipSet:
-// { <cid1> <cid2> <cid3> }
-func (ts TipSet) String() string {
-	return ts.ToSortedCidSet().String()
+// At returns the i'th block in the tipset.
+// An index outside the half-open range [0, Len()) will panic.
+func (ts TipSet) At(i int) *Block {
+	return ts.blocks[i]
 }
 
-// Equals returns true if the tipset contains the same blocks as another set.
-// Equality is not tested deeply.  If blocks of two tipsets are stored at
-// different memory addresses but have the same cids the tipsets will be equal.
-func (ts TipSet) Equals(ts2 TipSet) bool {
-	return ts.ToSortedCidSet().Equals(ts2.ToSortedCidSet())
+// IsSolo tests whether the tipset has a single block.
+func (ts TipSet) IsSolo() bool {
+	return len(ts.blocks) == 1
 }
 
-// ToSortedCidSet returns a SortedCidSet containing the Cids in the
-// TipSet.
+// ToSortedCidSet returns a SortedCidSet containing the CIDs in the tipset.
 func (ts TipSet) ToSortedCidSet() SortedCidSet {
 	s := SortedCidSet{}
-	for _, b := range ts {
+	for _, b := range ts.blocks {
 		s.Add(b.Cid())
 	}
 	return s
 }
 
-// ToSlice returns the slice of *Block containing the tipset's blocks.
-// Sorted.
+// ToSlice returns an ordered slice of pointers to the tipset's blocks.
 func (ts TipSet) ToSlice() []*Block {
-	sl := make([]*Block, len(ts))
-	var i int
-	for it := ts.ToSortedCidSet().Iter(); !it.Complete(); it.Next() {
-		sl[i] = ts[it.Value()]
-		i++
-	}
-	return sl
+	slice := make([]*Block, len(ts.blocks))
+	copy(slice, ts.blocks)
+	return slice
 }
 
-// MinTicket returns the smallest ticket of all blocks in the tipset.
+// MinTicket returns the smallest ticket of all blocks in the tipset, and nil error.
 func (ts TipSet) MinTicket() (Signature, error) {
-	if len(ts) == 0 {
-		return nil, ErrEmptyTipSet
-	}
-	blks := ts.ToSlice()
-	min := blks[0].Ticket
-	for i := range blks[0:] {
-		if bytes.Compare(blks[i].Ticket, min) < 0 {
-			min = blks[i].Ticket
-		}
-	}
-	return min, nil
+	return ts.blocks[0].Ticket, nil
 }
 
-// Height returns the height of a tipset.
+// Height returns the height of a tipset, and nil error.
 func (ts TipSet) Height() (uint64, error) {
-	if len(ts) == 0 {
-		return uint64(0), ErrEmptyTipSet
-	}
-	return uint64(ts.ToSlice()[0].Height), nil
+	return uint64(ts.blocks[0].Height), nil
```

Comment:
I like how making TipSet immutable simplifies the error handling here and elsewhere.

Reply:
It's stated at the struct level. I've asked in a Go forum for advice.